zkevm-circuits/src/witness/tx.rs

```diff
@@ -307,52 +318,448 @@ impl Transaction {
             })
             .collect()
     }
-}
 
-impl Encodable for Transaction {
-    fn rlp_append(&self, s: &mut RlpStream) {
-        s.begin_list(9);
-        s.append(&Word::from(self.nonce));
-        s.append(&self.gas_price);
-        s.append(&Word::from(self.gas));
-        if let Some(addr) = self.callee_address {
-            s.append(&addr);
+    pub(crate) fn gen_rlp_witness<F: Field>(
+        &self,
+        is_hash: bool,
+        challenges: &Challenges<Value<F>>,
+    ) -> Vec<RlpFsmWitnessRow<F>> {
+        let (rlp_bytes, format) = if is_hash {
+            (
+                self.rlp_signed.clone(),
+                match self.tx_type {
+                    TxTypes::Eip155 => TxHashEip155,
+                    TxTypes::PreEip155 => TxHashPreEip155,
+                    TxTypes::Eip1559 => TxHashEip1559,
+                    TxTypes::L1Msg => L1MsgHash,
+                    _ => unimplemented!(),
+                },
+            )
         } else {
-            s.append(&"");
+            (
+                self.rlp_unsigned.clone(),
+                match self.tx_type {
+                    TxTypes::Eip155 => TxSignEip155,
+                    TxTypes::PreEip155 => TxSignPreEip155,
+                    TxTypes::Eip1559 => TxSignEip1559,
+                    _ => unimplemented!(),
+                },
+            )
+        };
+
+        let tx_id = self.id as u64;
+        let mut witness = vec![];
+        let rom_table = format.rom_table_rows();
+        let keccak_rand = challenges.keccak_input();
+        let word_rand = challenges.evm_word();
+        let rlp_bytes_rlc = rlp_bytes
+            .iter()
+            .scan(Value::known(F::zero()), |rlc, &byte| {
+                *rlc = *rlc * keccak_rand + Value::known(F::from(byte as u64));
+
+                Some(*rlc)
+            })
+            .collect::<Vec<_>>();
+        let mut cur = SmState {
+            tag: rom_table[0].tag,
+            state: DecodeTagStart,
+            tag_idx: 0,
+            tag_length: 0,
+            tag_value_acc: Value::known(F::zero()),
+            byte_idx: 0,
+            depth: 0,
+        };
+        // TODO: what's the reason that cur_rom_row is of type `vector`.
+        // At the beginning of parsing tag we actually do not know the next tag
+        // only after we finished parsing the current tag, then we can know
+        let mut cur_rom_row = vec![0];
+        let mut remaining_bytes = vec![rlp_bytes.len()];
+        let mut witness_table_idx = 0;
+
+        // This map keeps track
+        // - the last row in the witness table of each parsed tag,
+        // - the row in the rom table of each parsed tag.
+        // And this map is used to fill the tag_next column in the witness table
+        let mut tag_rom_row_map = BTreeMap::new();
+        let mut is_output;
+        let mut is_none;
+        let mut rlp_tag;
+        let mut lb_len = 0;
+
+        loop {
```

Comment:
It might be a good idea to refactor this loop in a later follow-up PR, especially the `match cur.state` part, so that each "current state" is handled by a function, just to make reviewing easier. As of now, I don't really see any mistakes in the witness generation logic though :)

Reply:
Also, in general I would prefer replacing `if..else` with `match` so that its more readable. But that's a minor point for now.